### PR TITLE
refactor model to accomodate WOF, fix past issues and add more tests

### PR DESCRIPTION
**Pull request by missinglink**
a big refactor to accommodate the WOF taxonomy, while still being backwards compatible with the Quattroshapes taxonomy.

the javascript API for pelias/model is itself not backwards compatible, so when import pipelines upgrade, they will need to change their usage of this library, the specifics are outlined below.

changes:

**constructor** - has 3 parameters now instead of 2, they correspond to `source`, `layer` and `source_id`:
```javascript
// old
var poi = new Document( 'osmnode', 1001 )

// new
var poi = new Document( 'osm', 'node', 1001 )
```

**parent getter/setter** - introduced `addParent` and `removeParent`.
```javascript
var poi = new Document( 'osm', 'node', 1001 )
      .setAdmin( 'admin0', 'Great Britain' )
      .setAdmin( 'neighborhood', 'Shoreditch' )
      .addParent( 'country', 'Great Britain', '1001', 'United Kingdom' )
      .addParent( 'neighbourhood', 'Shoreditch', '2002' )
```

**[edit]** the API was changed, originally I proposed `setParent()` and `getParent()` but have now changed it to `addParent()` and `removeParent()` to enable multiple hierarchies (internally it's stored as an array).

**[edit]** the API was changed again, calls to `setParent()` and `getParent()` accept multiple arguments `( $field, $name, $id, $abbr )` as above, the last arg $abbr is optional.

**[edit]** removed `removeParent`

**note:** confusingly you need to use `neighborhood` (US spelling) for Quattroshapes and `neighbourhood` (UK spelling) for WOF. getting it wrong will throw an Error, as will using any invalid property name.

**more getters/setters** - introduced `*Source`, `*Layer`, `*SourceId`.
```javascript
var poi = new Document( 'osm', 'node', 1001 )
      .getSource() // osm
      .getLayer() // node
      .getSourceId() // 1001
```

**automatically clones doc.name to doc.phrase** - in the past each controller had to manually copy `doc.name` to `doc.phrase` such as [this](https://github.com/pelias/openstreetmap/blob/65860f335e66698f150323e373d213710cc8e542/stream/dbmapper.js#L12-L15), you don't need to do this any more, it's done [for you](https://github.com/pelias/model/pull/21/files#diff-b6601ead27f936beeb24aab59f789fd3R33).

**changes to doc.getGid()**
```javascript
var poi = new Document( 'osm', 'node', 1001 )
      .getGid() // 'osm:node:1001'
```

**change doc.boundaries to doc.shape** - the name of the field was wrong, the javascript API remains the same, so nothing to change there, you can use it as before:
```javascript
var poi = new Document( 'osm', 'node', 1001 )
      .setPolygon( geojsonObject /* any valid geojson object */ )
```

**tests** - added a bunch more tests, including a serialization test which shows how the document is serialized to JSON form before being sent to elasticsearch, this is worth a read:
https://github.com/pelias/model/blob/wof_refactor/test/serialize/test.js

**[edit]** added waaay more tests

**backwards compatibility** - the API is fully backwards compatible *but* requires you to set **both** quattroshapes `.setAdmin()` fields *and* WOF `.addParent()` fields if you would like to add both to the index.

**[edit]** the API was changed, `alpha3` is now *only* a root level element, you cannot set `alpha3` via `.addParent('alpha3','XXX')`.

**do not assume that the model is doing any format shifting or has any logic regarding either dataset beyond which property names are valid.**

### CODE REVIEW

**Comment by trescube** [comment]
What we decided to do:

- modify addParent to always have a value for abbreviation (requires ES research)
- remove removeParent
- move region code lookup to wof-admin-lookup
- modify label generation to consider arrays of names

In the end, we'll have internally consistent array sizes with possible data duplication in the case of multiple hierarchies.

File: Document.js [hunk 1]
```diff
@@ -5,21 +5,40 @@ var pkg = require('./package'),
...
+
+  // set the elasticsearch '_type' property to be the same as $layer
+  // this may be removed/modified in the future if required but will mean
+  // that; for example; all 'address' data ends up in the same '_type', even
+  // if it comes from different sources.
+  this.setType( layer );
 }
 
+Document.prototype.toJSON = function(){
+  this.phrase = this.name;
```
Comment (orangejulius) [inline Document.js]: Does this line imply that there's a side effect to calling `.toJSON()` on a Document?

File: Document.js [hunk 2]
```diff
@@ -168,12 +236,12 @@ Document.prototype.setCentroid = function( centroid ){
 };
 Document.prototype.getCentroid = model.get( 'center_point' );
 
-// boundaries
-Document.prototype.setPolygon = model.set( 'boundaries' )
+// shape
```
Comment (orangejulius) [inline Document.js]: this is bridge
http://abstrusegoose.com/432

File: test/document/polygon.js [hunk 3]
```diff
@@ -5,23 +5,23 @@ module.exports.tests = {};
...
   });
 };
 
 module.exports.tests.setPolygon = function(test) {
   test('setPolygon', function(t) {
-    var doc = new Document('mytype','myid');
-    t.equal(doc.boundaries, undefined, 'polygon undefined');
+    var doc = new Document('mysource','mylayer','myid');
+    t.equal(doc.shape, undefined, 'polygon undefined');
     t.equal(doc.setPolygon({ 'foo': 'bar' }), doc, 'chainable');
```
Comment (orangejulius) [inline test/document/polygon.js]: what about testing a real polygon here? (could be done as a separate PR since this PR didn't write this test).

File: test/serialize/fixtures.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+      'properties': {},
+      'geometry': {
+        'type': 'Polygon',
+        'coordinates': [
+          [
+            [ 163.82, -47.75 ],
+            [ 163.82, -33.90 ],
+            [ 179.42, -33.90 ],
+            [ 179.42, -47.75 ],
+            [ 163.82, -47.75 ]
```
Comment (orangejulius) [inline test/serialize/fixtures.js]: This must win an award for the most simplified geometry on Earth

**Comment by orangejulius** [comment]
I just realized that the model and schema as it stands at this moment will potentially cause problems with the `/place` endpoint.

TL;DR: we might find ourselves in another situation where we can't unambiguously refer to individual documents with the current setup of using layer as `_type`

## Suggested pre-reading
* https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-multi-get.html

## How this PR as it stands stores source/layer

* source: put in a newly created field called `source`
* layer: put into the `_type` field

The `_type` field which is a special Elasticsearch field that allows you to do cool things. Most notably. `_id` and `_type` are the only two document fields that the Multi Get endpoint allows you to query on.

## How the place endpoint queries for multiple documents

It uses the Multi-Get api to query for many `_type`/`_id` combos.

## How is this a problem?

layer + id is, theoretically, not enough to uniquely distinguish documents in our index. However at current i can't think of any actual ambiguities:

* openaddress records are all type address, but with a simple integer id. OSM adddress records will have an id like `node:1234` or `way:1234`
* geonames venues have integer ids, OSM venues again have node or way as a prefix
* Quattroshapes and WOF don't share any layers.

## Can we use the Elasticsearch [Bulk API](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html)?

Nope. It only supports insert/delete

## Can we just do bunch of individual queries in a for loop?

Yes but that would be really slow.

## Can we create a single elasticsearch query that returns a few records given their source, layer, and id?

Possibly but it doesn't sound fun. It would presumably have at least some performance impact since the Multi-Get interface is designed to be simpler than a full query.

**Comment by dianashk** [comment]
:+1:

Let's address the unique id issue separately. After discussing with the team, we agreed that what we have in place currently will work, as long as each importer adds unique info to the `_id` field to avoid conflicts. We need to keep `_type` simple because it's what we use to filter by layers and we want to avoid having to do the _fancy-mapping-dance_ to associate a bunch of types with `address` for example.
